x-pack/plugin/migrate/src/main/java/org/elasticsearch/xpack/migrate/action/CopyLifecycleIndexMetadataTransportAction.java

```diff
@@ -64,7 +63,8 @@ public CopyLifecycleIndexMetadataTransportAction(
         this.executor = new SimpleBatchedAckListenerTaskExecutor<>() {
             @Override
             public Tuple<ClusterState, ClusterStateAckListener> executeTask(UpdateIndexMetadataTask task, ClusterState clusterState) {
-                return new Tuple<>(applyUpdate(clusterState, task), task);
+                final ProjectMetadata projectMetadata = projectResolver.getProjectMetadata(clusterState);
+                return new Tuple<>(applyUpdate(clusterState, projectMetadata, task), task);
```

Comment:
I understand why you're passing both a cluster state and a project metadata, but that's something we generally avoid. We use a `ProjectState` if we need access to the cluster state _and_ are doing something in the scope of a specific project. But in this case, I'd prefer making `applyUpdate` return a `ProjectMetadata` and building the new cluster state here in `executeTask`. That way we keep the scope of `applyUpdate` clear in that it only operates on a single project (and we avoid constructing a `ProjectState`). But I'll let it up to you which one you choose.

Reply:
Same, this should be handled in 7d17f7a0ea6e464fe37012e2259acb35bce091a9, _I think_.